polkadot/xcm/src/lib.rs

```diff
@@ -195,12 +210,26 @@ macro_rules! versioned_type {
 				}
 			}
 		}
-		impl MaxEncodedLen for $n {
-			fn max_encoded_len() -> usize {
-				<$v3>::max_encoded_len()
+		// internal macro that handles some edge cases
+		versioned_type!(@internal $n, $v3, $v4, $v5, $($($gen),+)?);
```

Comment:
It doesn't help readability that having a generic implies not implementing `MaxEncodedLen`.

You could also match `VersionedXcm` specifically and not implement `MaxEncodedLen` in this case. Or have another optional input in the macro: `@do_not_derive_max_encoded_len`.

```suggestion
		// We don't derive `MaxEncodedLen` for type with generic.
		versioned_type!(@maybe_max_encoded_len $n, $v3, $v4, $v5, $($($gen),+)?);
```